File: EstructurasPython/stack.py
from node import Node

class Stack:
    def __init__(self):
        self.top = None
        self.size = 0

    def push(self, data):
        node = Node(data)

        if self.top:
            node.next = self.top
            self.top = node
        else:
            self.top = node

        self.size += 1
    def pop(self):
        if self.top:
            data = self.top.data
            self.size -= 1
            if self.top.next:
                self.top = self.top.next
            else:
                self.top = None
            return data
        else:
            return "The stack is empty"    
        
    def peek(self):
        if self.top:
            return self.top.data
        else:
            return "The stack is empty"
        
    def clear(self):
        while self.top:
            self.pop()
            
    def recursiveReverse(self):
        if self.top:
            store_data= self.top.data
            self.pop()
            self.recursiveReverse()
            self.push(store_data)
            return self.__str__
        else:
            return None



    def __str__(self):
        string = "["
        current = self.top
        for i in range(self.size):
            string += str(current)
            if i != self.size-1:
                string += str(", ")
            current = current.next
        string += "]"

        return string


if __name__=="__main__":
    food = Stack()
    food.push("pizza")
    food.push("hamburger")
    food.push("Ham")
    food.push("Hot dog")
    print(food)
    food.pop()
    print(food)
    print(food.peek())
    print(food.recursiveReverse())

File: EstructurasJavascript/stack/stack.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Stack {
    constructor(){
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }
    peek(){
        return this.top;
    }
    push(value){
        const newNode = new Node(value);
        if(this.length === 0){
            this.top = newNode;
            this.bottom = newNode;
        }else{
            const holdingPointer = this.top;
            this.top = newNode;
            this.top.next = holdingPointer;
        }
        this.length++;
        return this
    }
    pop(){
        if (!this.top){
            return null;
        }
        if(this.top === this.bottom){
            this.bottom = null;
        }
        let current = this.top
        this.top = current.next;
        this.length--;
        return this;
    }
    

}

let MyStack = new Stack()
MyStack.push('google')
MyStack.push('udemy')
console.log(MyStack.peek(), MyStack.length);
MyStack.push('discord')
console.log(MyStack.peek(), MyStack.length);
MyStack.pop()
console.log(MyStack.peek(), MyStack.length);
